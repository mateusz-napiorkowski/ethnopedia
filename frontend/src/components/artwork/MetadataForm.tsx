import React, { useState, ChangeEvent, useEffect } from 'react';
import { ReactComponent as PlusIcon } from "../../assets/icons/plus.svg"
import { ReactComponent as MinusIcon } from "../../assets/icons/minus.svg";
import FormField from './FormField';
import { Category } from '../../@types/Category';


// let example_data: Category[] = [
//   { name: "Tytuł", values: ["tytuł utworu"], subcategories: [] },
//   { name: "Wykonawca", values: ["nazwa zespołu"], subcategories: [
//     { name: "Wykonawca nr 1", values: ["Imię Nazwisko 1"] },
//     { name: "Wykonawca nr 2", values: ["Imię Nazwisko 2"] } ] },
//   { name: "Region", values: ["Wielkopolska"], subcategories: [
//     { name: "Podregion", values: ["WielkopQolska Północna"] },
//     { name: "Podregion etnograficzny", values: ["Szamotulskie"] },
//     { name: "Powiat", values: ["Szamotulski"]} ] }
// ]

function transformToNewStructure(data: Category[], collectionName: string): { categories: Category[]; collectionName: string } {
  return { categories: data, collectionName: collectionName };
}



interface MetadataFormProps {
  initialFormData: Category[];
  setDataToInsert: any
}


const MetadataForm: React.FC<MetadataFormProps> = ({ initialFormData, setDataToInsert }) => {
  const [formDataList, setFormDataList] = React.useState<Category[]>(initialFormData); // Ustaw początkowe dane

  const createOrEdit = window.location.href.split("/")[window.location.href.split("/").length-1]
  let collectionName = window.location.href.split("/")[window.location.href.split("/").length-2]
  if(createOrEdit === "edit-artwork") {
    collectionName = window.location.href.split("/")[window.location.href.split("/").length-4]
  }

  useEffect(() => {
    setDataToInsert(transformToNewStructure(formDataList, collectionName))
  }, [formDataList, collectionName, setDataToInsert])

  useEffect(() => {
    const handleResize = () => {
      setFormDataList((prev) => [...prev]); // Wymuszenie ponownego renderowania
    };
    window.addEventListener('resize', handleResize);
    return () => window.removeEventListener('resize', handleResize);
  }, []);


  const [jsonOutput, setJsonOutput] = useState<string>('');
  const handleShowJson = () => {
    let jsonData = transformToNewStructure(formDataList, collectionName);
    setJsonOutput(JSON.stringify(jsonData, null, 2));
  };

  const handleInputChange = (index: string, e: ChangeEvent<HTMLInputElement>) => {
    const { name, value } = e.target;
    const indexParts = index.split('-').map(Number);  // e.g. index '2-0-1' -> [2, 0, 1]

    setFormDataList((prevDataList) =>
        updateNestedValue(prevDataList, indexParts, name, value)
    );
  };

  // recursively update the value of a nested object
  const updateNestedValue = (dataList: Category[], indexParts: number[], name: string, value: string): Category[] => {
    if (indexParts.length === 0) {
      return dataList;
    }

    const [currentIndex, ...remainingIndexParts] = indexParts;

    return dataList.map((item, i) => {
      if (i === currentIndex) {
        if (remainingIndexParts.length === 0) {
          if (name === 'values') {
            // If it is a value field
            return { ...item, [name]: [value] };
          } else {
            // If it is a name field
            return { ...item, [name]: value };
          }
        } else {
          // We need to go deeper, recurse
          const updatedSubcategories = updateNestedValue(item.subcategories || [], remainingIndexParts, name, value);
          return { ...item, subcategories: updatedSubcategories };
        }
      } else {
        return item;
      }
    });
  };

  const handleAddCategory = () => {
    setFormDataList((prevDataList) => [...prevDataList, { name: '', values: [''], subcategories: [] }]);
  };

  const handleAddSubcategory = (index: string) => {
    const indexParts = index.split('-').map(Number);  // e.g. index '2-0-1' -> [2, 0, 1]

    setFormDataList((prevDataList) =>
        addSubcategory(prevDataList, indexParts)
    );
  };

  const addSubcategory = (dataList: Category[], indexParts: number[]): Category[] => {
    if (indexParts.length === 0) {
      return [...dataList, { name: '', values: [''], subcategories: [] }];
    }

    const [currentIndex, ...remainingIndexParts] = indexParts;

    return dataList.map((item, i) => {
      if (i === currentIndex) {
        if (remainingIndexParts.length === 0) {
          // We're at the correct item, add a subcategory
          return {
            ...item,
            subcategories: [
              ...(item.subcategories || []),
              { name: '', values: [''] }
            ]
          };
        } else {
          // We need to go deeper, recurse
          const updatedSubcategories = addSubcategory(item.subcategories || [], remainingIndexParts);
          return { ...item, subcategories: updatedSubcategories };
        }
      } else {
        return item;
      }
    });
  };


  const handleRemove = (index: string) => {
    const indexParts = index.split('-').map(Number);  // e.g. index '2-0-1' -> [2, 0, 1]

    setFormDataList((prevDataList) =>
        removeCategory(prevDataList, indexParts)
    );
  };

  const removeCategory = (dataList: Category[], indexParts: number[]): Category[] => {
    if (indexParts.length === 0) {
      return dataList;
    }

    const [currentIndex, ...remainingIndexParts] = indexParts;

    return dataList.map((item, i) => {
      if (i === currentIndex) {
        if (remainingIndexParts.length === 0) {
          // We're at the correct item, remove it
          return null;
        } else {
          // We need to go deeper, recurse
          const updatedSubcategories = removeCategory(item.subcategories || [], remainingIndexParts);
          return { ...item, subcategories: updatedSubcategories };
        }
      } else {
        return item;
      }
    }).filter(item => item !== null) as Category[];
  };

  return (
      // <div style={{ overflowY: 'auto', height: '70vh', minWidth: '2000px'}}> {/* TODO */}
        <div style={{ overflowY: 'auto'}}> {/* TODO */}
          <form>
            {formDataList.map((formData, index) => (
                <FormField
                    key={index.toString()}
                    index={index.toString()}
                    level={0}
                    formData={formData}
                    formDataList={formDataList}
                    handleInputChange={handleInputChange}
                    handleRemove={handleRemove}
                    handleAddSubcategory={handleAddSubcategory}
                />
            ))}
            <div className="flex flex-col justify-between items-start mt-1 space-y-4">
              <button type="button" onClick={handleAddCategory} title="Dodaj kategorię">
                <PlusIcon />
              </button>
              {/*<button type="button" onClick={handleShowJson}>*/}
              {/*  Show JSON*/}
              {/*</button>*/}
            </div>
          </form>
          <pre>{jsonOutput}</pre>
      </div>
  );
}

export default MetadataForm;
