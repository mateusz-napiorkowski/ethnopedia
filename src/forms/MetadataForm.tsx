import { Formik, Field, Form } from "formik"

const MetadataForm = () => {
    const ipcRenderer = (window as any).electron.ipcRenderer

    const onFormSubmit = (e: any) => {
        console.log("onFormSubmit")
        ipcRenderer.send("submit:metadataForm", "form-submit")
    }

    return <>
        <h1>MetadataInfoForm</h1>

        <Formik
            initialValues={{ email: "", password: "" }}
            onSubmit={(values, { setSubmitting }) => {
                onFormSubmit("form-submit")
            }}>
            <Form>
                <div>
                    <label htmlFor="email">Email:</label>
                    <Field type="email" name="email" />
                </div>
                <div>
                    <label htmlFor="password">Password:</label>
                    <Field type="password" name="password" />
                </div>
                <button type="submit">Submit</button>
            </Form>
        </Formik>
    </>
}
export default MetadataForm