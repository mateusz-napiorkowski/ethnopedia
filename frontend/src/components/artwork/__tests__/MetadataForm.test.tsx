import '@testing-library/jest-dom';
import { render } from '@testing-library/react'
import MetadataForm from "../MetadataForm"
import { MemoryRouter, Routes, Route } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { QueryClient, QueryClientProvider } from 'react-query';

const queryClient = new QueryClient();
const user = userEvent.setup()

const collectionData = {
    "_id": "675ddf8c1e6d01766fbc5b2e",
    "name": "example collection",
    "description": "example collection description",
    "categories": [
        {name: 'Tytuł', subcategories: [{name: "Podtytuł", subcategories: []}]},
        {name: 'Artyści', subcategories: []},
        {name: 'Rok', subcategories: []},
    ],
    "__v": 0
}
const artworkId = "6752ddca46e3ca48231024dc"

const treeData = [
    {
        "name": "Tytuł",
        "value": "Tytuł utworu",
        "subcategories": [
            {
                "name": "Podtytuł",
                "value": "Podtytuł utworu",
                "subcategories": []
            }
        ]
    },
    {
        "name": "Artyści",
        "value": "Nazwa Artysty",
        "subcategories": []
    },
    {
        "name": "Rok",
        "value": "1999",
        "subcategories": []
    }
]

const renderComponent = (edit = false) => {
        return render(
            <QueryClientProvider client={queryClient}>
                <MemoryRouter initialEntries={[
                    edit ? 
                        `/collections/${collectionData._id}/artworks/${artworkId}/edit-artwork`
                        : `/collections/${collectionData._id}/create-artwork/`
                    ]}>
                    <Routes>
                        <Route path={
                            edit ?
                                "/collections/:collectionId/artworks/:artworkId/edit-artwork"
                                : "/collections/:collectionId/create-artwork/"
                        } element={<MetadataForm initialMetadataTree={edit ? treeData : undefined} categoryPaths={edit ? undefined : ['Tytuł', 'Tytuł.Podtytuł', 'Artyści', 'Rok']} setFieldValue={() => {}}/>}/>
                    </Routes>  
                </MemoryRouter>
            </QueryClientProvider>    
        );
};

describe("EmptyCollectionMessage", () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    it("should render component correctly when in create mode", () => {
        const {container} = renderComponent()
        expect(container).toMatchSnapshot()
    })

    it("should render component correctly when in edit mode", () => {
        const {container} = renderComponent(true)
        expect(container).toMatchSnapshot()
    })

    it("should fill and clear form fields correctly when focusing on input fields by clicking them", async () => {
        const {container, getByLabelText} = renderComponent()
        
        const titleInputField = getByLabelText("Tytuł-input")
        const subtitleInputField = getByLabelText("Podtytuł-input")
        const artistsInputField = getByLabelText("Artyści-input")
        const yearInputField = getByLabelText("Rok-input")

        await user.type(subtitleInputField, "an artwork subtitle")
        await user.type(titleInputField, "an artwork title")
        await user.type(artistsInputField, "artist name")
        await user.clear(titleInputField)
        await user.type(yearInputField, "199")

        expect(container).toMatchSnapshot()
    })

    it("should fill and clear form fields correctly when focusing on input fields by using keyboard keys", async () => {
        const {container, getByLabelText} = renderComponent()
        
        const titleInputField = getByLabelText("Tytuł-input")

        await user.click(titleInputField)
        await user.keyboard(
            "{t}{y}{t}{u}{ł}{ }{u}{t}{w}{o}{r}{u}" +
            "{enter}" +
            "{p}{o}{d}{t}{y}{t}{u}{ł}{ }{u}" +
            "{arrowdown}" +
            "{n}{a}{z}{w}{a}{ }{a}{r}{t}{y}{s}{t}{y}" +
            "{arrowup}" +
            "{t}{w}{o}{r}{u}" +
            "{tab}{tab}" +
            "{1}{9}{9}{9}"
        );
        expect(container).toMatchSnapshot()
    })

    it("should stop filling category input field when string length exceeds 100 characters", async () => {
        const {getByLabelText} = renderComponent()
        
        const titleInputField = getByLabelText("Tytuł-input")
        const veryLongString = "a".repeat(150)

        await user.type(titleInputField, veryLongString)
        
        expect(titleInputField).toHaveValue("a".repeat(100))
    })
})