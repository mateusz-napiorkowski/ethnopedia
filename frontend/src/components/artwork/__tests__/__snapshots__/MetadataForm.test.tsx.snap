// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmptyCollectionMessage should render component correctly when in create mode 1`] = `
<div>
  <div
    data-testid="category-tree"
  >
    <div
      class="relative "
    >
      <div
        class="relative w-full mb-2"
      >
        <label
          class="absolute left-2 top-2 text-xs text-gray-500 dark:text-gray-400 bg-white px-1 z-10 dark:bg-gray-800 max-w-[calc(100%-16px)] truncate"
          for="field-0"
          title="Tytuł"
        >
          Tytuł
        </label>
        <input
          aria-label="Tytuł-input"
          autocomplete="off"
          class="block w-full rounded-none border border-gray-300 bg-white px-3 pt-7 pb-2 text-sm
               placeholder-transparent focus:border-blue-400 focus:outline-none
               focus:shadow-[inset_0_0_3px_rgba(59,130,246,0.3)]
               dark:border-gray-600 dark:bg-gray-800 dark:text-white dark:placeholder-gray-500
               dark:focus:border-blue-400 dark:focus:shadow-[inset_0_0_3px_rgba(59,130,246,0.3)]"
          id="field-0"
          maxlength="250"
          type="text"
          value=""
        />
      </div>
      <div
        class="ml-4 relative"
      >
        <div
          class="relative pl-6"
        >
          <div
            class="absolute left-0 w-px bg-gray-300 dark:bg-gray-600"
            style="top: -0.5rem; bottom: 2.1rem;"
          />
          <div
            class="absolute left-0 top-6 h-px w-6 bg-gray-300 dark:bg-gray-600"
          />
          <div
            class="relative w-full mb-2"
          >
            <label
              class="absolute left-2 top-2 text-xs text-gray-500 dark:text-gray-400 bg-white px-1 z-10 dark:bg-gray-800 max-w-[calc(100%-16px)] truncate"
              for="field-0-0"
              title="Podtytuł"
            >
              Podtytuł
            </label>
            <input
              aria-label="Podtytuł-input"
              autocomplete="off"
              class="block w-full rounded-none border border-gray-300 bg-white px-3 pt-7 pb-2 text-sm
               placeholder-transparent focus:border-blue-400 focus:outline-none
               focus:shadow-[inset_0_0_3px_rgba(59,130,246,0.3)]
               dark:border-gray-600 dark:bg-gray-800 dark:text-white dark:placeholder-gray-500
               dark:focus:border-blue-400 dark:focus:shadow-[inset_0_0_3px_rgba(59,130,246,0.3)]"
              id="field-0-0"
              maxlength="250"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="relative "
    >
      <div
        class="relative w-full mb-2"
      >
        <label
          class="absolute left-2 top-2 text-xs text-gray-500 dark:text-gray-400 bg-white px-1 z-10 dark:bg-gray-800 max-w-[calc(100%-16px)] truncate"
          for="field-1"
          title="Artyści"
        >
          Artyści
        </label>
        <input
          aria-label="Artyści-input"
          autocomplete="off"
          class="block w-full rounded-none border border-gray-300 bg-white px-3 pt-7 pb-2 text-sm
               placeholder-transparent focus:border-blue-400 focus:outline-none
               focus:shadow-[inset_0_0_3px_rgba(59,130,246,0.3)]
               dark:border-gray-600 dark:bg-gray-800 dark:text-white dark:placeholder-gray-500
               dark:focus:border-blue-400 dark:focus:shadow-[inset_0_0_3px_rgba(59,130,246,0.3)]"
          id="field-1"
          maxlength="250"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="relative "
    >
      <div
        class="relative w-full mb-2"
      >
        <label
          class="absolute left-2 top-2 text-xs text-gray-500 dark:text-gray-400 bg-white px-1 z-10 dark:bg-gray-800 max-w-[calc(100%-16px)] truncate"
          for="field-2"
          title="Rok"
        >
          Rok
        </label>
        <input
          aria-label="Rok-input"
          autocomplete="off"
          class="block w-full rounded-none border border-gray-300 bg-white px-3 pt-7 pb-2 text-sm
               placeholder-transparent focus:border-blue-400 focus:outline-none
               focus:shadow-[inset_0_0_3px_rgba(59,130,246,0.3)]
               dark:border-gray-600 dark:bg-gray-800 dark:text-white dark:placeholder-gray-500
               dark:focus:border-blue-400 dark:focus:shadow-[inset_0_0_3px_rgba(59,130,246,0.3)]"
          id="field-2"
          maxlength="250"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <div>
    <label
      class="block text-sm font-bold text-gray-700 dark:text-white my-2"
      for="dropzone-file"
    >
      Wgraj pliki
    </label>
    <p
      class="block text-sm font-normal"
    >
      Obsługiwane formaty plików: mei, midi, musicxml, xml, wav, mp3, txt.
    </p>
    <p
      class="block text-sm font-normal mb-2"
    >
      Maksymalny rozmiar pliku: 
      <span
        class=""
      >
        25 MB.
      </span>
    </p>
    <label
      aria-label="upload"
      class="flex flex-col items-start justify-start p-2 border-2 border-gray-200
                                    border-dashed rounded-lg cursor-pointer bg-gray-50 dark:hover:bg-gray-600
                                    dark:bg-gray-800 hover:bg-gray-100 dark:border-gray-600
                                    dark:hover:border-gray-500 dark:hover:bg-gray-700"
      for="dropzone-file"
    >
      <div
        class="flex flex-row items-center justify-center gap-4"
      >
        <svg
          class="w-12 h-12 text-gray-500 dark:text-gray-400"
        />
        <p
          class="text-sm font-normal text-gray-500 dark:text-gray-400"
        >
          Kliknij, aby przesłać pierwszy plik
        </p>
      </div>
      <input
        class="hidden"
        id="dropzone-file"
        type="file"
      />
    </label>
  </div>
</div>
`;

exports[`EmptyCollectionMessage should render component correctly when in edit mode 1`] = `
<div>
  <div
    data-testid="category-tree"
  >
    <div
      class="relative "
    >
      <div
        class="relative w-full mb-2"
      >
        <label
          class="absolute left-2 top-2 text-xs text-gray-500 dark:text-gray-400 bg-white px-1 z-10 dark:bg-gray-800 max-w-[calc(100%-16px)] truncate"
          for="field-0"
          title="Tytuł"
        >
          Tytuł
        </label>
        <input
          aria-label="Tytuł-input"
          autocomplete="off"
          class="block w-full rounded-none border border-gray-300 bg-white px-3 pt-7 pb-2 text-sm
               placeholder-transparent focus:border-blue-400 focus:outline-none
               focus:shadow-[inset_0_0_3px_rgba(59,130,246,0.3)]
               dark:border-gray-600 dark:bg-gray-800 dark:text-white dark:placeholder-gray-500
               dark:focus:border-blue-400 dark:focus:shadow-[inset_0_0_3px_rgba(59,130,246,0.3)]"
          id="field-0"
          maxlength="250"
          type="text"
          value="Tytuł utworu"
        />
      </div>
      <div
        class="ml-4 relative"
      >
        <div
          class="relative pl-6"
        >
          <div
            class="absolute left-0 w-px bg-gray-300 dark:bg-gray-600"
            style="top: -0.5rem; bottom: 2.1rem;"
          />
          <div
            class="absolute left-0 top-6 h-px w-6 bg-gray-300 dark:bg-gray-600"
          />
          <div
            class="relative w-full mb-2"
          >
            <label
              class="absolute left-2 top-2 text-xs text-gray-500 dark:text-gray-400 bg-white px-1 z-10 dark:bg-gray-800 max-w-[calc(100%-16px)] truncate"
              for="field-0-0"
              title="Podtytuł"
            >
              Podtytuł
            </label>
            <input
              aria-label="Podtytuł-input"
              autocomplete="off"
              class="block w-full rounded-none border border-gray-300 bg-white px-3 pt-7 pb-2 text-sm
               placeholder-transparent focus:border-blue-400 focus:outline-none
               focus:shadow-[inset_0_0_3px_rgba(59,130,246,0.3)]
               dark:border-gray-600 dark:bg-gray-800 dark:text-white dark:placeholder-gray-500
               dark:focus:border-blue-400 dark:focus:shadow-[inset_0_0_3px_rgba(59,130,246,0.3)]"
              id="field-0-0"
              maxlength="250"
              type="text"
              value="Podtytuł utworu"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="relative "
    >
      <div
        class="relative w-full mb-2"
      >
        <label
          class="absolute left-2 top-2 text-xs text-gray-500 dark:text-gray-400 bg-white px-1 z-10 dark:bg-gray-800 max-w-[calc(100%-16px)] truncate"
          for="field-1"
          title="Artyści"
        >
          Artyści
        </label>
        <input
          aria-label="Artyści-input"
          autocomplete="off"
          class="block w-full rounded-none border border-gray-300 bg-white px-3 pt-7 pb-2 text-sm
               placeholder-transparent focus:border-blue-400 focus:outline-none
               focus:shadow-[inset_0_0_3px_rgba(59,130,246,0.3)]
               dark:border-gray-600 dark:bg-gray-800 dark:text-white dark:placeholder-gray-500
               dark:focus:border-blue-400 dark:focus:shadow-[inset_0_0_3px_rgba(59,130,246,0.3)]"
          id="field-1"
          maxlength="250"
          type="text"
          value="Nazwa Artysty"
        />
      </div>
    </div>
    <div
      class="relative "
    >
      <div
        class="relative w-full mb-2"
      >
        <label
          class="absolute left-2 top-2 text-xs text-gray-500 dark:text-gray-400 bg-white px-1 z-10 dark:bg-gray-800 max-w-[calc(100%-16px)] truncate"
          for="field-2"
          title="Rok"
        >
          Rok
        </label>
        <input
          aria-label="Rok-input"
          autocomplete="off"
          class="block w-full rounded-none border border-gray-300 bg-white px-3 pt-7 pb-2 text-sm
               placeholder-transparent focus:border-blue-400 focus:outline-none
               focus:shadow-[inset_0_0_3px_rgba(59,130,246,0.3)]
               dark:border-gray-600 dark:bg-gray-800 dark:text-white dark:placeholder-gray-500
               dark:focus:border-blue-400 dark:focus:shadow-[inset_0_0_3px_rgba(59,130,246,0.3)]"
          id="field-2"
          maxlength="250"
          type="text"
          value="1999"
        />
      </div>
    </div>
  </div>
  <div>
    <label
      class="block text-sm font-bold text-gray-700 dark:text-white my-2"
      for="dropzone-file"
    >
      Wgraj pliki
    </label>
    <p
      class="block text-sm font-normal"
    >
      Obsługiwane formaty plików: mei, midi, musicxml, xml, wav, mp3, txt.
    </p>
    <p
      class="block text-sm font-normal mb-2"
    >
      Maksymalny rozmiar pliku: 
      <span
        class=""
      >
        25 MB.
      </span>
    </p>
    <label
      aria-label="upload"
      class="flex flex-col items-start justify-start p-2 border-2 border-gray-200
                                    border-dashed rounded-lg cursor-pointer bg-gray-50 dark:hover:bg-gray-600
                                    dark:bg-gray-800 hover:bg-gray-100 dark:border-gray-600
                                    dark:hover:border-gray-500 dark:hover:bg-gray-700"
      for="dropzone-file"
    >
      <div
        class="flex flex-row items-center justify-center gap-4"
      >
        <svg
          class="w-12 h-12 text-gray-500 dark:text-gray-400"
        />
        <p
          class="text-sm font-normal text-gray-500 dark:text-gray-400"
        >
          Kliknij, aby przesłać pierwszy plik
        </p>
      </div>
      <input
        class="hidden"
        id="dropzone-file"
        type="file"
      />
    </label>
  </div>
</div>
`;
